src/hotspot/share/ci/ciInstanceKlass.cpp

```diff
@@ -534,6 +534,43 @@ int ciInstanceKlass::compute_nonstatic_fields() {
   return fields->length();
 }
 
+
+// Generic routine to handle synthetic field population for both static
+// and non-static base multifields.
+ciField* ciInstanceKlass::populate_synthetic_multifields(ciField* field) {
+  ASSERT_IN_VM;
+  if (!field->is_multifield_base()) {
+    return field;
+  }
+  int sec_fields_count = 0;
+  ciMultiField* mfield = static_cast<ciMultiField*>(field);
+  Arena* arena = CURRENT_ENV->arena();
+  InstanceKlass* k = get_instanceKlass();
+
+  for (JavaFieldStream fs(k); !fs.done(); fs.next()) {
+    if (fs.offset() == field->offset_in_bytes()) {
+      assert(mfield == field, "Duplicate multifield for a given offset");
+      fieldDescriptor& fd = fs.field_descriptor();
+      assert(fd.is_multifield_base(), "");
+      mfield = new (arena) ciMultiField(&fd);
+      sec_fields_count = fd.secondary_fields_count(fd.index()) - 1;
+      mfield->add_secondary_fields(new (arena) GrowableArray<ciField*>(arena, sec_fields_count, 0, NULL));
+      if (sec_fields_count == 0) {
+        break;
+      }
+    } else if (sec_fields_count && fs.is_multifield()) {
+      fieldDescriptor& fd = fs.field_descriptor();
+      mfield->secondary_fields()->append(new (arena) ciField(&fd));
+      sec_fields_count--;
+      if (sec_fields_count == 0) {
+        break;
+      }
+    }
+  }
+
+  return mfield;
+}
```

Comment:
This whole logic is quite similar with the one in `compute_nonstatic_fields_impl` when the field is a multifield. But the code style is not the same. For example, we check the `sec_fields_count` when field is multifield. This is right to me, in case other multifield is appended to a previous multifield_base which is not its expected one, although we do not have such cases now. But this similar check is missing in `compute_nonstatic_fields_impl`, isn't it? 

Another suggestion is: can we directly use `ciMultiField` instead of `ciField` as the method argument type? Besides, may I ask why we have to new another `ciMultiField`, set the secondary_fields to it and return? Can we modify the input field instead and change the method's return type to `void`?

Reply:
Yes, _compute_nonstatic_fields_impl_ is only called from [ InlineTypeNode constructor](https://github.com/openjdk/valhalla/blob/lworld%2Bvector/src/hotspot/share/opto/inlinetypenode.hpp#L40), but while parsing field access bytecodes "withfield/getfield/putfield" both c1 and c2 compiler uses _ciBytecodeStream::get_field() API_,  since it directly operates over bytecode hence does not take into account the effect of @multifield annotation over it and earlier used to create only one ciField for base multifield instead of creating a hierarchical ciMutifield structure.

In addition, ClassFiledParser creates[ contiguous sequence of FieldInfo ](https://github.com/openjdk/valhalla/blob/lworld%2Bvector/src/hotspot/share/classfile/classFileParser.cpp#L1623) for both base and subsequent synthetic multifield, hense it's not possible to associate synthetic multifields with any other base multifield.

>>> But this similar check is missing in compute_nonstatic_fields_impl, isn't it?

We can add an assertion to check that secondary fields count always match the number of synthetic FieldInfos after base multifield.